### PR TITLE
Remove SPARC backend

### PR DESCRIPTION
**Pull request by mshinwell**
The SPARC backend, which only generates 32-bit code, is difficult to test these days due to lack of available hardware.  It is suspected that there is no actual use for this backend in the wild at the moment.  The maintenance burden of maintaining this backend would seem to be higher than any extra cost (due to bitrot) that its removal would incur in the future should it be resurrected and ported to modern SPARC64 hardware.

### CODE REVIEW

**Comment by johnwhitington** [comment]
I have about half a dozen customers of my PDF software using this backend, the most recent of whom purchased from us in 2015.

Now, since I'm not able to contribute to its maintenance, or willing to pay for someone to contribute to it, I can hardly complain. But it is still in use.

**Comment by mshinwell** [comment]
Hmm.  We may need to keep it around in that case.  Do you know if the target is Solaris or Linux?

**Comment by avsm** [comment]
If there's an actual user (thanks @johnwhitington!) then I can dig out SparcStation hardware in the Computer Lab.  One issue is that Linux kernel support for sparc32 has lagged for years, so it's likely that an old Solaris installation is the only practical OS target to support.  I don't have a copy of Solaris lying around...

**Comment by johnwhitington** [comment]
Solaris 8 in the most recent case. The customer compiled it themselves.

Because of the following issue I know OCaml builds ok on solaris SPARC at least up to 4.01.0:

http://caml.inria.fr/mantis/view.php?id=6214

In this instance, I compiled it by remote login to the customer's machine.

@avsm Thanks, but I think I still have access to sparc hardware via polarhome.com, if I can dig out the login details, so I'm willing to test build before 4.04.

**Comment by xavierleroy** [comment]
@johnwhitington : I have a polarhome.com account as well, but recently they reduced disk quotas so much that you can't even check out the OCaml source tree, let alone build it...

**Comment by johnwhitington** [comment]
@xavierleroy Oh dear. I remember having to disable camlp4 last time (when it was in-tree), because there wasn't enough disk space to do the build, but I didn't realise it had got so bad.

**Comment by alainfrisch** [comment]
(Closing the PR, since it seems this backend is not going away soon.  Feel free to continue using this thread to discuss how to best access such hardware for testing purposes.)

**Comment by DemiMarie** [comment]
Could we cross-compile from a different architecture, and they run the binaries?

**Comment by DemiMarie** [comment]
This is a motivation to provide a backend – even an unoptimized one – that generates C code and/or LLVM IR, to provide platform independence.

**Comment by avsm** [comment]
> This is a motivation to provide a backend – even an unoptimized one – that generates C code and/or LLVM IR, to provide platform independence.

The bytecode `ocamlc` compiler is a (mostly) unoptimized and highly portable backend that only needs a C compiler.

**Comment by bobzhang** [comment]
hi @xavierleroy , we have solaris sparc machine, we don't use it in production for ocaml code base, but I can help set up a jenkins job, where should I report failures?
CC @shayne-fletcher

**Comment by const-rs** [comment]
I have SPARC/Solaris machine (last OpenSolaris). Current ocaml d6c72fc82a8950851da5f121b80a4e1c8bff377a builds and works well enough for me.

The only problem I had - OPAM compiled with 4.02 (its lib-ext does not compile with current trunk) crashed several times with SIGSEGV somewhere deep in GC. Though I can not reproduce it now (perhaps it was a problem with installation).

**Comment by damiendoligez** [comment]
@bobzhang I can see a few options:
- set it up as a slave for Inria's CI system (needs you to accept ssh connections from Inria)
- set it up as a Travis slave (if that's possible)
- set it up in your own Jenkins and send reports to ocaml-ci-notifications(at)inria.fr, which is the address used by Inria's CI system. BTW I can give you access to the Inria CI system if you want.

**Comment by bobzhang** [comment]
@damiendoligez I will try to set up a Jenkins job this week and see what can be improved

**Comment by mshinwell** [comment]
Reopening this pull request.  Having a backend which the core team has no means of testing seems untenable, roughly speaking.

@damiendoligez @bobzhang Have there been any updates on trying to get one of these machines in the CI?  I think it needs to be running Solaris as per the discussion above.

**Comment by mshinwell** [comment]
I checked with @johnwhitington and the conclusion was that it was ok to remove the SPARC backend now.  The code concerns compiles with versions of OCaml going back some way, and in any case, the bytecode version could always be used.

If there are no further objections soon then I'll rebase this patch for the backend's removal.

**Comment by damiendoligez** [comment]
@mshinwell as far as I know, the Sparc CI was never live.

**Comment by xavierleroy** [comment]
All right, it is clear now that no one wants to pay the (financial and human) costs of proper SPARC/Solaris support in OCaml.  I'm in favor of removing this compiler back-end.

**Comment by mshinwell** [comment]
This is ready for review.

File: asmrun/signals_asm.c [hunk 1]
```diff
@@ -158,21 +158,9 @@ int caml_set_signal_action(int signo, int action)
 /* Machine- and OS-dependent handling of bound check trap */
 
 #if defined(TARGET_power) \
-  || defined(TARGET_s390x) \
-  || (defined(TARGET_sparc) && defined(SYS_solaris))
```
Comment (xavierleroy) [inline asmrun/signals_asm.c:162]: Explain why you're touching s390x support, here?

File: configure [hunk 2]
```diff
@@ -696,7 +696,7 @@ esac
 
 case "$target" in
   # PR#5088: autodetection is unreliable on ARM.  PR#5280: also on MIPS.
-  sparc*-*-*|hppa*-*-*|arm*-*-*|mips*-*-*)
```
Comment (xavierleroy) [inline configure]: Don't remove, SPARC is still supported in bytecode, and this alignment business affects the bytecode system.

File: otherlibs/num/.depend [hunk 3]
```diff
@@ -6,7 +6,6 @@ bng_arm64.$(O): bng_arm64.c
 bng_digit.$(O): bng_digit.c
 bng_ia32.$(O): bng_ia32.c
 bng_ppc.$(O): bng_ppc.c
-bng_sparc.$(O): bng_sparc.c
```
Comment (xavierleroy) [inline otherlibs/num/.depend]: Don't remove, this can still help with the bytecode version of the library

File: otherlibs/num/bng_sparc.c [hunk 4]
```diff
@@ -1,77 +0,0 @@
-/**************************************************************************/
```
Comment (xavierleroy) [inline otherlibs/num/bng_sparc.c]: More generally, don't touch otherlibs/num at all

**Review by xavierleroy — CHANGES_REQUESTED** [review]
You're removing too much stuff, see below.  Remember, this is just native-code SPARC support that is being removed, not bytecode support.

File: configure (hunk 2 above)
Comment (mshinwell) [reply configure]: Ah yes, sorry, I misread that as being for native only.

File: asmrun/signals_asm.c (hunk 1 above)
Comment (mshinwell) [reply asmrun/signals_asm.c:162]: I'm not -- the s390x line is still there, just without the backslash...

File: otherlibs/num/.depend (hunk 3 above)
Comment (mshinwell) [reply otherlibs/num/.depend]: Resurrected

File: otherlibs/num/bng_sparc.c (hunk 4 above)
Comment (mshinwell) [reply otherlibs/num/bng_sparc.c]: Likewise

**Review by xavierleroy — APPROVED** [review]

**Comment by xavierleroy** [comment]
Looks good to me.  Merging in trunk.

**Comment by glaubitz** [comment]
> The SPARC backend, which only generates 32-bit code, is difficult to test these days due to lack of available hardware. It is suspected that there is no actual use for this backend in the wild at the moment. The maintenance burden of maintaining this backend would seem to be higher than any extra cost (due to bitrot) that its removal would incur in the future should it be resurrected and ported to modern SPARC64 hardware.

I'm super late to the party, but in case anyone is interested, Oracle engineers actually ported the SPARC backend to 64 bits. The patch can be found [here](https://github.com/sparclinux/ocaml/commit/6125e690f956da35914ae514a204ae662ca33059).

If someone is interested on hacking on 64-bit SPARC, there are two fast SPARC servers running both Debian unstable as well as Solaris 11.4 available in the [GCC Compile Farm](https://gcc.gnu.org/wiki/CompileFarm).

**Pull request merged**